cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc

```diff
@@ -240,7 +240,7 @@ struct Month {
 // ----------------------------------------------------------------------
 // Extract day from temporal types
 
-template <typename Duration, typename Localizer>
+template <typename Duration, typename Localizer, typename = void*>
```

Comment:
Add a third type parameter to distinguish `DurationType` from others. Defaults to `void *` so the existing `TemporalComponentExtractBase`s and other kernels that don't have `DurationType` overloads don't have to change.